modules/solace/solace.go

```diff
@@ -0,0 +1,291 @@
+func (s *SolaceContainer) WithEnv(env map[string]string) *SolaceContainer {
+	if s.envVars == nil {
+		s.envVars = make(map[string]string)
+	}
+	for k, v := range env {
+		s.envVars[k] = v
+	}
+	return s
+}
+
+// WithExposedPorts allows adding extra exposed ports
+func (s *SolaceContainer) WithExposedPorts(ports ...string) *SolaceContainer {
+	s.exposedPorts = append(s.exposedPorts, ports...)
+	return s
+}
+
+const (
+	DefaultImage = "solace/solace-pubsub-standard:latest"
+	DefaultVpn   = "default"
+	DefaultUser  = "root"
+	DefaultPass  = "password"
+)
+
+func NewSolaceContainer(ctx context.Context, image string) *SolaceContainer {
+	if image == "" {
+		image = DefaultImage
+	}
+	return &SolaceContainer{
+		Username: DefaultUser,
+		Password: DefaultPass,
+		Vpn:      DefaultVpn,
+		Image:    image,
+		ctx:      ctx,
+	}
+}
+
+func (s *SolaceContainer) WithShmSize(size int64) *SolaceContainer {
+	s.shmSize = size
+	return s
+}
+
+// waitForSolaceActive waits for the Solace broker to be fully active by checking the system log
+func (s *SolaceContainer) waitForSolaceActive(ctx context.Context) error {
+	const maxAttempts = 60
+	const intervalSeconds = 1
+
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		// Execute grep command to check for "Primary Virtual Router is now active" in system log
+		code, _, err := s.Container.Exec(ctx, []string{"grep", "-R", "Primary Virtual Router is now active", "/usr/sw/jail/logs/system.log"})
+
+		if err == nil && code == 0 {
+			// Found the message, broker is active
+			return nil
+		}
+
+		// Wait before next attempt
+		time.Sleep(intervalSeconds * time.Second)
+	}
+
+	return fmt.Errorf("timeout waiting for Solace broker to become active after %d seconds", maxAttempts)
+}
+
+func (s *SolaceContainer) Run(ctx context.Context) error {
+	shmSize := s.shmSize
+	if shmSize == 0 {
+		shmSize = 1 << 30 // Default 1GB
+	}
+
+	// Merge env vars
+	env := map[string]string{
+		"username_admin_globalaccesslevel": "admin",
+		"username_admin_password":          "admin",
+	}
+	for k, v := range s.envVars {
+		env[k] = v
+	}
+
+	// Collect exposed ports from all known services
+	defaultServices := []Service{ServiceAMQP, ServiceManagement, ServiceSMF, ServiceREST, ServiceMQTT}
+	portsMap := make(map[string]struct{})
+	var ports []string
+	for _, svc := range defaultServices {
+		portStr := fmt.Sprintf("%d/tcp", svc.Port)
+		if _, exists := portsMap[portStr]; !exists {
+			ports = append(ports, portStr)
+			portsMap[portStr] = struct{}{}
+		}
+	}
+	if len(s.exposedPorts) > 0 {
+		ports = append(ports, s.exposedPorts...)
+	}
+
+	// --- CLI script generation for queue/topic config ---
+	// Reference: https://docs.solace.com/Admin-Ref/CLI-Reference/VMR_CLI_Commands.html
+	var cliScript string
+	if len(s.queues) > 0 {
+		cliScript += "enable\nconfigure\n"
+
+		// Create VPN if not default
+		if s.Vpn != DefaultVpn {
+			cliScript += fmt.Sprintf("create message-vpn %s\n", s.Vpn)
```

Comment:
suggestion: use a multi-line string with backticks, maybe? Or a template file that is copied. See redpanda module for a similar approach.

Reply:
addressed that, moved into a separate file with explicit methods